nixos/modules/services/networking/vpp.nix

```diff
@@ -0,0 +1,525 @@
+          defaultText = lib.literalExpression ''pkgs.writeText "vpp-\${name}-startup.conf" (toString config.services.vpp.instances.\${name}.startupConfig);'';
+          description = ''
+            File to run as a script on startup in {command}`vppctl`, passed
+            to VPP's `startup-config` option in the `unix` section.
+
+            Setting this option will override {option}`startupConfig`.
+          '';
+        };
+      };
+    };
+in
+{
+  options.services.vpp = {
+    hugepages = {
+      autoSetup = mkOption {
+        default = false;
+        example = true;
+        description = ''
+          Whether to automatically setup hugepages for use with FD.io's Vector Packet Processor.
+
+          If any programs other than VPP use hugepages or you want to use 1GB hugepages, it's recommended
+          to keep this `false` and set them up manually using {option}`boot.kernel.sysctl` or {option}`boot.kernelParams`.
+        '';
+      };
+
+      count = mkOption {
+        type = types.ints.positive;
+        default = 1024;
+        example = 512;
+        description = "Number of 2MB hugepages to setup on the system if {option}`services.vpp.hugepages.autoSetup` is enabled.";
+      };
+    };
+
+    instances = mkOption {
+      default = { };
+      type = types.attrsOf (types.submodule vppOpts);
+      description = ''
+        VPP supports multiple instances for testing or other purposes.
+        If you don't require multiple instances of VPP you can define just the one.
+      '';
+      example = lib.literalExpression ''
+        {
+          "main" = {
+            enable = true;
+            group = "vpp-main";
+            settings = { };
+          };
+          "test" = {
+            enable = false;
+            group = "vpp-test";
+            settings = { };
+          };
+        };
+      '';
+    };
+  };
+
+  config =
+    let
+      #TODO: systemd hardening
+      mkInstanceServiceConfig = instance: {
+        description = "Vector Packet Processing Process";
+        after = [
+          "syslog.target"
+          "network.target"
+          "auditd.service"
+        ];
+        serviceConfig = {
+          ExecStartPre =
+            [
+              "-${pkgs.coreutils}/bin/rm -f /dev/shm/db /dev/shm/global_vm /dev/shm/vpe-api"
+            ]
+            ++ (lib.optional (
+              instance.kernelModule != null
+            ) "-/run/current-system/sw/bin/modprobe ${instance.kernelModule}");
+          ExecStart = "${instance.package}/bin/vpp -c ${instance.settingsFile}";
+          Type = "simple";
+          Restart = "on-failure";
+          RestartSec = "5s";
+          RuntimeDirectory = "vpp";
+        };
+        wantedBy = [ "multi-user.target" ];
+      };
+      instances = lib.attrValues cfg.instances;
+    in
+    lib.mkIf (lib.any (v: v.enable) instances) {
+      environment.systemPackages = with pkgs; [
```

Comment:
Same as elsewhere, I would do 
```nix
environment.systemPackages = [ pkgs.vpp ];
```

Reply:
Additionally, would it make sense to use `cfg.package` instead of `pkgs.vpp`?

Reply:
Nice idea with `cfg.package`, unfortunately it's not feasible here since the option is instance-specific. Removed the `with` though.